lib/units/api/controllers/user.js

```diff
@@ -0,0 +1,332 @@
+var util = require('util')
+
+var Promise = require('bluebird')
+var _ = require('lodash')
+
+var dbapi = require('../../../db/api')
+var logger = require('../../../util/logger')
+var datautil = require('../../../util/datautil')
+var wire = require('../../../wire')
+var wireutil = require('../../../wire/util')
+
+var log = logger.createLogger('api:controllers:user')
+
+module.exports = {
+  getUser: getUser
+, getUserDevices: getUserDevices
+, addUserDevice: addUserDevice
+, getUserDeviceBySerial: getUserDeviceBySerial
+, deleteUserDeviceBySerial: deleteUserDeviceBySerial
+, remoteConnectUserDeviceBySerial: remoteConnectUserDeviceBySerial
+, remoteDisconnectUserDeviceBySerial: remoteDisconnectUserDeviceBySerial
+, getUserAccessTokens: getUserAccessTokens
+}
+
+function getUser(req, res) {
+  res.json({
+    success: true
+  , user: req.user
+  })
+}
+
+function getUserDevices(req, res) {
+  var fields = req.swagger.params.fields.value
+
+  dbapi.loadUserDevices(req.user.email)
+    .then(function(cursor) {
+      return Promise.promisify(cursor.toArray, cursor)()
+        .then(function(list) {
+          var deviceList = []
+
+          list.forEach(function(device) {
+            datautil.normalize(device, req.user)
+            if (fields) {
+              device = _.pick(device, fields.split(','))
+            }
+            deviceList.push(device)
+          })
+
+          res.json({
+            success: true
+          , devices: deviceList
+          })
+        })
+    })
+    .catch(function(err) {
+      log.error('Failed to load device list: ', err.stack)
+      res.status(500).json({
+        success: false
+      })
+    })
+}
+
+function getUserDeviceBySerial(req, res) {
+  var serial = req.swagger.params.serial.value
+  var fields = req.swagger.params.fields.value
+
+  dbapi.loadDevice(serial)
+    .then(function(device) {
+      if (device) {
+        datautil.normalize(device, req.user)
+
+        if (device.owner && device.owner.email === req.user.email) {
+          if(fields) {
+            device = _.pick(device, fields.split(','))
+          }
+
+          res.json({
+            success: true
+          , device: device
+          })
+        }
+        else {
+          res.status(401).json({
+            success: false
+          , description: 'Device is not owned by you'
+          })
+        }
+      }
+      else {
+        res.status(404).json({
+          success: false
+        , description: 'Device not found'
+        })
+      }
+    })
+    .catch(function(err) {
+      log.error('Failed to load device "%s": ', req.params.serial, err.stack)
+      res.status(500).json({
+        success: false
+      })
+    })
+}
+
+function addUserDevice(req, res) {
+  var serial = req.body.serial
+  var timeout = req.body.timeout || null
+
+  dbapi.loadDevice(serial)
+    .then(function(device) {
+      if (device) {
```

Comment:
If possible, try to have these the other way around. It's kind of difficult to see where the error handling is since the function body is pretty long. Then you keep reading the function and suddenly you see the error handling code, and have to try to recall or scroll up to see what it was even checking. So this is basically an overall statement about the codebase, but I would prefer `if (!device) { return res.status(404, ..) } .. long function body.. ` to `if (device) { .. long function body .. } else { res.status(404, ..) }`, unless the function is ~10 lines max. I'm sure we have existing bad examples somewhere in the codebase already, but it just makes the code a bit more difficult to review since the flow is disturbed.